File: scripts/make_clean_h5ad.R


source('/home/kodermam/packages/scprocess/scripts/utils.R')

make_clean_sces(
sel_b         = 'GSM8612262_SRR31256771', 
sel_run       = 'GSM8612262_SRR31256771', 
integration_f = '/pstore/data/mus-brain-analysis/studies/test_project/output/test_integration/integrated_dt_test_project_2025-01-01.csv.gz', 
h5_paths_f    = '/pstore/data/mus-brain-analysis/studies/test_project/output/test_hvg/hvg_paths_test_project_2025-01-01.csv', 
coldata_f     = '/pstore/data/mus-brain-analysis/studies/test_project/output/test_qc/coldata_dt_all_cells_test_project_2025-01-01.csv.gz',
run_var       = 'sample_id',
batch_var     = 'sample_id',
clean_sce_f   = '/pstore/data/mus-brain-analysis/studies/test_project/output/test_integration/test_sce_cells_clean_GSM8612262_SRR31256771_test_project_2025-01-01.rds',
rowdata_f     = '/pstore/data/mus-brain-analysis/studies/test_project/output/test_qc/rowdata_dt_test_project_2025-01-01.csv.gz')


make_clean_sces <- function(sel_b, sel_run, integration_f, h5_paths_f, 
  coldata_f, rowdata_f, run_var, batch_var, clean_sce_f) {
  # load, exclude doublets
  int_dt        = fread(integration_f) %>%
    .[ is_dbl == FALSE & in_dbl_cl == FALSE ]
  
  # check if ok
  ok_batches    = int_dt[[batch_var]] %>% unique
  if (!(sel_b %in% ok_batches)) {
    message('excluded ', sel_b, '; creating empty sce file.')
    file.create(clean_sce_f)
    return(NULL)
  }
  message('creating clean sce file for ', sel_b)

  # get some subsets
  message('  getting values specific to ', sel_b)
  batch_int   = int_dt %>% .[ get(batch_var) == sel_b ]
  batch_ids   = batch_int$cell_id
  batch_cols  = fread(coldata_f) %>% setkey("cell_id") %>% .[ batch_ids ]

  # get sce object
  message('  loading counts into sce')
  h5_paths    = fread(h5_paths_f)
  filtered_f  = h5_paths[ get(run_var) == sel_run ]$amb_filt_f %>% unique
  assert_that(file.exists(filtered_f))
  sce         = .get_sce_clean(filtered_f, sel_run, run_var, rowdata_f,  subset_cells = batch_ids)
  
  # add things to colData
  message('  adding coldata')
  sce         = .add_coldata(sce, batch_cols)
  message('  adding integration outputs')
  sce         = .add_int_variables(sce, batch_int)

  # save
  message('  saving sce')
  saveRDS(sce, clean_sce_f, compress = FALSE)

  message('done!')
}


.get_sce_clean <- function(filtered_f, sel_run, run_var, rowdata_f, subset_cells = batch_ids){
  # read matrix
  mat         = .get_h5_mx(filtered_f, paste0(sel_run, ':')) %>% .sum_SUA
  if (!is.null(subset_cells)) {
    message('    subsetting sce to specified cells')
    assert_that( all(subset_cells %in% colnames(mat)) )
    mat         = mat[, subset_cells]
  }

  # read rowdata
  rows_dt       = fread(rowdata_f) %>% setkey('ensembl_id')
  keep_ids      = rows_dt$ensembl_id
  assert_that(all(keep_ids %in% rownames(mat)))
  mat           = mat[keep_ids, ]
  rows_dt       = rows_dt[ rownames(mat) ]
  assert_that( identical(rownames(mat), rows_dt$ensembl_id) )
  rownames(mat) = rows_dt$gene_id
  
  # make sce object
  sce_clean           = SingleCellExperiment(assays = list(counts = mat), rowData = rows_dt)
  sce_clean$cell_id   = colnames(mat)
  rownames(sce_clean) = rowData(sce_clean)$gene_id
  
  # convert to TsparseMatrix
  counts(sce_clean) = counts(sce_clean) %>% as("TsparseMatrix")

  return(sce_clean)
}

.add_coldata <- function(sce, batch_cols) {
  # some checks
  assert_that( colnames(colData(sce)) == "cell_id" )
  assert_that( all(colnames(sce) == batch_cols$cell_id) )

  # add columns
  cols_df       = batch_cols %>% as("DataFrame") %>% set_rownames(colnames(sce))
  colData(sce)  = cols_df

  return(sce)
}

.add_int_variables <- function(sce_clean, batch_int) {
  assert_that( all(colnames(sce_clean) == batch_int$cell_id) )
  # get useful integration variables, add to sce object
  int_vs      = c('UMAP1', 'UMAP2', str_subset(names(batch_int), "RNA_snn_res"))
  for (v in int_vs) {
    if (str_detect(v, "RNA_snn_res")) {
      colData(sce_clean)[[ v ]] = batch_int[[ v ]] %>% factor
    } else {
      colData(sce_clean)[[ v ]] = batch_int[[ v ]]
    }
  }

  return(sce_clean)
}

File: scripts/make_clean_h5ad.py
import polars as pl
import anndata as ad
import scanpy as sc
from scipy.sparse import csc_matrix
import os
import h5py
import argparse
import re
import numpy as np


def make_clean_h5ad(sel_b, sel_run, integration_f, h5_paths_f, 
  coldata_f, rowdata_f, run_var, batch_var, clean_h5ad_f):
  
  # load, exclude doublets
  int_dt = pl.read_csv(integration_f)
  int_dt = int_dt.filter((pl.col("is_dbl") == False) & (pl.col("in_dbl_cl") == False))
    
  # check if ok
  ok_batches = int_dt[batch_var].unique().to_list()
  if sel_b not in ok_batches:
    print(f'excluded {sel_b}; creating empty file.')
    open(clean_h5ad_f, 'a').close()
    return None
    
  print(f'creating clean h5ad file for {sel_b}')

  # get some subsets
  print(f'  getting values specific to {sel_b}')
  batch_int  = int_dt.filter(pl.col(batch_var) == sel_b)
  batch_ids  = batch_int["cell_id"].to_list()

  # get adata objects
  print('  loading counts into AnnData')
  h5_paths   = pl.read_csv(h5_paths_f)
  filtered_f = h5_paths.filter(pl.col(run_var) == sel_run).select("amb_filt_f").unique().item()
  assert os.path.exists(filtered_f), f"File {filtered_f} not found"
  
  adata = _get_clean_adata(filtered_f, sel_run, rowdata_f, coldata_f, batch_ids)
  
  # add integration outputs
  print('  adding integration variables')
  adata = _add_int_variables(adata, batch_int)
  
  # save
  print('  saving h5ad')
  
  # check for strange types that can't be saved
  obj_cols = adata.obs.select_dtypes(include=['object']).columns
  for col in obj_cols:
    adata.obs[col] = adata.obs[col].fillna("").astype(str)

  cat_cols = adata.obs.select_dtypes(include=['category']).columns
  for col in cat_cols:
    if adata.obs[col].isnull().any():
      if "" not in adata.obs[col].cat.categories:
        adata.obs[col] = adata.obs[col].cat.add_categories("")
      adata.obs[col] = adata.obs[col].fillna("")

  adata.write_h5ad(clean_h5ad_f)
  print('done!')



def _get_clean_adata(filtered_f, sel_run, rowdata_f, coldata_f, sub_cols ):
  
  with h5py.File(filtered_f, 'r') as f:
    indptr      = f['matrix/indptr'][:]
    indices     = f['matrix/indices'][:]
    data        = f['matrix/data'][:]
    features    = f['matrix/features/name'][:]
    barcodes    = f['matrix/barcodes'][:].astype('U16')
    num_rows    = f['matrix/shape'][0]
    num_cols    = f['matrix/shape'][1]

  # make a csc sparse matrix
  sua_csc_mat = csc_matrix((data, indices, indptr), shape=(num_rows, num_cols))
  csc_mat, uniq_features = sum_SUA(sua_csc_mat, features)
  
  barcodes    = barcodes.astype('<U21')
  barcodes    = [f"{sel_run}:{bc}" for bc in barcodes]  
  barcodes    = np.array(barcodes)

  # create anndata object
  adata = ad.AnnData(X=csc_mat.T)
  adata.obs_names = barcodes
  adata.var_names = uniq_features
  
  # filter features and add feature metadata
  feats_dt = pl.read_csv(rowdata_f)
  keep_ids = feats_dt["ensembl_id"].to_list()
  adata    = adata[:, adata.var_names.isin(keep_ids)].copy()

  var_df = (pl.DataFrame({"ensembl_id": adata.var_names})
    .join(feats_dt, on="ensembl_id", how="left")
  )
  adata.var = var_df.to_pandas().set_index("gene_id")
  
  # filter cells and add cell metadata
  batch_cols = pl.read_csv(coldata_f).filter(pl.col("cell_id").is_in(sub_cols))
  adata      = adata[adata.obs_names.isin(sub_cols), :].copy()

  obs_df = (pl.DataFrame({"cell_id": adata.obs_names})
    .join(batch_cols, on = "cell_id", how = 'left')
  )
  adata.obs = obs_df.to_pandas().set_index('cell_id')
  
  return adata


def _add_int_variables(adata, batch_int):
  # make sure things are aligned
  batch_int_pd = batch_int.to_pandas().set_index("cell_id")
  batch_int_pd = batch_int_pd.reindex(adata.obs_names)

  # find umap and clustering cols
  int_vs = ["UMAP1", "UMAP2"] + [col for col in batch_int_pd.columns if "RNA_snn_res" in col]
    
  for v in int_vs:
    if "RNA_snn_res" in v:
      # convert clusters to categorical
      adata.obs[v] = batch_int_pd[v].astype('category')
    else:
      adata.obs[v] = batch_int_pd[v]

  return adata


# same function as in hvgs.py
def sum_SUA(sua_mat, row_names):
  # define some objects
  row_names = row_names.astype(str)
  types = ['_S$', '_U$', '_A$']
  mats = []

  # sum up each type of gene mapping
  for t in types:
    indices = [i for i, name in enumerate(row_names) if re.search(t,name)]
    mat = sua_mat[indices, :]
    mats.append(mat)

  # give them names
  genes = [re.sub(r'_[SUA]$', '', name) for name in row_names]
  uniq_genes = list(dict.fromkeys(genes))
  uniq_genes = np.array(uniq_genes)

  # sum across the mappings
  mats_sum = sum(mats)

  return mats_sum, uniq_genes


if __name__ == "__main__":
  # get arguments
  parser  = argparse.ArgumentParser()
  parser.add_argument("sel_batch", type=str)
  parser.add_argument("sel_run", type=str)
  parser.add_argument("integration_f", type=str)
  parser.add_argument("h5_paths_f", type = str)
  parser.add_argument("coldata_f", type=str)
  parser.add_argument("rowdata_f", type=str)
  parser.add_argument("run_var", type = str)
  parser.add_argument("batch_var", type=str)
  parser.add_argument("clean_h5ad_f", type = str)

  # set up some locations
  args    = parser.parse_args()

  # run
  make_clean_h5ad(args.sel_batch, args.sel_run, args.integration_f, args.h5_paths_f, 
    args.coldata_f, args.rowdata_f, args.run_var, args.batch_var, args.clean_h5ad_f)